README.md
# CNav

A `custom navigation bar` a.k.a `CNav` with bubble click effect.<br>

[![pub package](https://img.shields.io/pub/v/cnav?style=for-the-badge)](https://pub.dev/packages/cnav) [![License: MIT](https://img.shields.io/badge/license-MIT-orange.svg?style=for-the-badge)](https://opensource.org/licenses/MIT) <a href="https://github.com/prodevplus/cnav/stargazers"><img alt="GitHub stars" src="https://img.shields.io/github/stars/prodevplus/cnav?style=for-the-badge"></a> <img alt="GitHub forks" src="https://img.shields.io/github/forks/prodevplus/cnav?style=for-the-badge"></a><a href="https://github.com/prodevplus/cnav/issues">  <img alt="GitHub issues" src="https://img.shields.io/github/issues/prodevplus/cnav?style=for-the-badge"></a> <a href="https://github.com/prodevplus/cnav/network">




## Overview

This project is inspired by this [post from Dribbble](https://dribbble.com/shots/7134849-Simple-Tab-Bar-Animation) and [The Boring Flutter Development Show, Ep. 35](https://www.youtube.com/watch?v=Qk_x9y_wGXQ)

This package gives you a cute bubble effect when you click on the navigation bar.

Dribbble:<br>
<img src="https://cdn.dribbble.com/users/2114584/screenshots/7134849/media/96e4a6002a476bad7bd809ac71e28698.gif" width="400">

Implemented:<br>
<img src="https://user-images.githubusercontent.com/13074740/190694868-8944e94d-dcf4-411a-87b1-90394cf9505d.gif" width="300">

## Gallery

<div style="text-align: center">
    <table>
        <tr>
            <td style="text-align: center">
                <a>
                    <img src="https://user-images.githubusercontent.com/13074740/190694257-6364b2ec-78aa-459b-b80a-72e9b3561c55.gif" width="200"/>
                </a>
            </td>            
            <td style="text-align: center">
                <a>
                    <img src="https://user-images.githubusercontent.com/13074740/190694542-e91fb7ed-91b3-4ef7-a22b-2dd642abe23e.gif" width="200"/>
                </a>
            </td>  
            <td style="text-align: center">
                <a>
                    <img src="https://user-images.githubusercontent.com/13074740/190694666-2ceab1ad-88dd-49b2-ad41-4b29b2f523d4.gif" width="200"/>
                </a>
            </td>  
        </tr>
        <tr>
            <td style="text-align: center">
                <a>
                    <img src="https://user-images.githubusercontent.com/13074740/190694689-b917abaa-f633-4bdd-8361-bae8d704fc3d.gif" width="200"/>
                </a>
            </td>  
            <td style="text-align: center">
                <a>
                    <img src="https://user-images.githubusercontent.com/13074740/190694708-07f02664-c112-4174-9dff-2c984db5d47e.gif" width="200"/>
                </a>
            </td>  
            <td style="text-align: center">
                <a>
                    <img src="https://user-images.githubusercontent.com/13074740/190694868-8944e94d-dcf4-411a-87b1-90394cf9505d.gif" width="200"/>
                </a>
            </td>  
        </tr>
        <tr>
            <td style="text-align: center">
                <a>
                    <img src="https://user-images.githubusercontent.com/13074740/190694884-2438257b-56f2-4600-b5d9-c4f3eeb3dc4f.gif" width="200"/>
                </a>
            </td>  
            <td style="text-align: center">
                <a>
                    <img src="https://user-images.githubusercontent.com/13074740/190694931-c3cb3a6b-9506-4b69-ad33-76b5a25b743c.gif" width="200"/>
                </a>
            </td>
            <td style="text-align: center">
                <a>
                    <img src="https://user-images.githubusercontent.com/13074740/190694987-5837e2e7-7d81-4c11-9c46-1ab78377b0cd.gif" width="200"/>
                </a>
            </td>
        </tr>
        <tr>
        <td style="text-align: center">
                <a>
                    <img src="https://user-images.githubusercontent.com/13074740/190695001-b083db20-cdbd-4893-955b-6d0d75b8d239.gif" width="200"/>
                </a>
            </td>  
            <td style="text-align: center">
                <a>
                    <img src="https://user-images.githubusercontent.com/13074740/190695011-547047f4-89e7-4627-84c2-2b70e6956944.gif" width="200"/>
                </a>
            </td>
        </tr>
    </table>
</div>

## How to install
Add this to your package's pubspec.yaml file:
```yaml
dependencies:
  cnav: <latest version>
```
## Documentation

```dart
CNavItem(
  {
    required this.icon,
    Widget selectedIcon,
    this.title,
    Text selectedTitle,
    this.badgeCount = 0,
    this.showBadge = false,
  })
    : selectedIcon = selectedIcon ?? icon,
      selectedTitle = selectedTitle ?? title;

  ///
  /// The icon of the item
  /// Typically the icon is an [Icon].
  ///

  final Widget icon;      
```


### You can customize these attributes in the navigation bar.<br>

| Attributes | Type | Description | Default |
| -------------------- | -------------- | ---------------- | --------------- |
| `scaleFactor` | `double`   | scale factor for the icon scale animation. | 0.2 |
| `elevation`   | `double` | The z-coordinate of this CNav | 8.0 |
| `items`   | `List` | item data in CNav  | required |
| `selectedColor`   | `Color` | [Color] when [CNavItem] is selected  | [blueAccent] |
| `unSelectedColor`   | `Color` | [Color] when [CNavItem] is not selected.  | grey[600] |
| `onTap`  | `Function(int)` | callback function when item tapped | null |
| `currentIndex`  | `int` | current index of navigation bar. | 0 |
| `iconSize`  | `double` | size of icon. also represent the max radius of bubble effect animation. | 24.0 |
| `backgroundColor`  | `Color` | Background color of [CNav] | Colors.white |
| `strokeColor`  | `Color` | stroke color | blueAccent |
| `bubbleCurve`  | `Curve` | animation curve of bubble effect | linear |
| `scaleCurve`  | `Curve` | animation curve of scale effect | linear |
| `borderRadius`  | `Radius` | border radius of navigation bar | Radius.zero |
| `isFloating`  | `bool` | control if CNav is floating | false |
| `blurEffect`  | `bool` | control if CNav show blur effect | false |
| `opacity`  | `double` | control CNav blur effect opacity when `blurEffect` is true | 0.8 |
| `margin`  | `EdgeInsetsGeometry` | Adds margin | EdgeInsets.zero |
| `padding`  | `EdgeInsetsGeometry` | Adds padding | EdgeInsets.zero |


**Attention**: If you set ```isFloating``` to  ```true```, I would recommend you to set ```extendBody``` to ```true``` in ```Scaffold``` for a better performance.

And for customize icon in the navigation bar, just put the icons you want in the ```CNavItem``` like this.
```dart
CNav(
        ...
        items: [
          CNavItem(
          icon: Icon(Icons.home),
          title: Text("hello"),
        ),
        CNavItem(
          icon: Icon(Icons.shopping_cart),
        ),
        CNavItem(
          icon: Icon(Icons.lightbulb_outline),
        ),
        CNavItem(
          icon: Icon(Icons.search),
        ),
        CNavItem(
          icon: Icon(Icons.account_circle),
        ),
        ],
        ...
      )
```

If you want add notification badge, just use like this
```dart
CNav(
        ...
        items: [
        CNavItem(
          icon: Icon(Icons.home),
          badgeCount: _badgeCounts[0],
          showBadge: _badgeShows[0],
        ),
        CNavItem(
          icon: Icon(Icons.shopping_bag),
          badgeCount: _badgeCounts[1],
          showBadge: _badgeShows[1],
        ),
        CNavItem(
          icon: Icon(Icons.lightbulb_outline),
          badgeCount: _badgeCounts[2],
          showBadge: _badgeShows[2],
        ),
        CNavItem(
          icon: Icon(Icons.search),
          badgeCount: _badgeCounts[3],
          showBadge: _badgeShows[3],
        ),
        CNavItem(
          icon: Icon(Icons.account_circle),
          badgeCount: _badgeCounts[4],
          showBadge: _badgeShows[4],
        ),
      ],
        ...
      )
```

To clear a badge, set `showBadge` to `false`

If you want add title under icon, just use like this
```dart
CNav(
      iconSize: 30.0,
      selectedColor: Color(0xff040307),
      strokeColor: Color(0x30040307),
      unSelectedColor: Color(0xffacacac),
      backgroundColor: Colors.white,
      items: [
        CNavItem(
          icon: Icon(Icons.home),
          title: Text("Home"),
        ),
        CNavItem(
          icon: Icon(Icons.shopping_cart),
          title: Text("Cart"),
        ),
        CNavItem(
          icon: Icon(Icons.lightbulb_outline),
          title: Text("Explore"),
        ),
        CNavItem(
          icon: Icon(Icons.search),
          title: Text("Search"),
        ),
        CNavItem(
          icon: Icon(Icons.account_circle),
          title: Text("Me"),
        ),
      ],
      currentIndex: _currentIndex,
      onTap: (index) {
        setState(() {
          _currentIndex = index;
        });
      },
    );
```

If you want use Svg, just use like this
```dart
CNav(
      iconSize: 30.0,
      margin: const EdgeInsets.all(20),
      borderRadius: const Radius.circular(10),
      strokeColor: const Color(0x300c18fb),
      backgroundColor: Colors.white,
      items: [
        CNavItem(
          title: const Text('Home'),
          icon: SvgPicture.asset(
            "assets/icons/svg/outline/home.svg",
            color: Colors.amber,
          ),
          selectedIcon: SvgPicture.asset(
            "assets/icons/svg/solid/home.svg",
            color: Colors.pinkAccent,
          ),
        ),
        CNavItem(
          title: const Text('Shopping'),
          icon: SvgPicture.asset(
            "assets/icons/svg/outline/shopping-bag.svg",
            color: Colors.amber,
          ),
          selectedIcon: SvgPicture.asset(
            "assets/icons/svg/solid/shopping-bag.svg",
            color: Colors.pinkAccent,
          ),
        ),
        CNavItem(
          title: const Text('Cloud'),
          icon: SvgPicture.asset(
            "assets/icons/svg/outline/cloud.svg",
            color: Colors.amber,
          ),
          selectedIcon: SvgPicture.asset(
            "assets/icons/svg/solid/cloud.svg",
            color: Colors.pinkAccent,
          ),
        ),
        CNavItem(
          title: const Text('Search'),
          icon: SvgPicture.asset(
            "assets/icons/svg/outline/magnifying-glass.svg",
            color: Colors.amber,
          ),
          selectedIcon: SvgPicture.asset(
            "assets/icons/svg/solid/magnifying-glass.svg",
            color: Colors.pinkAccent,
          ),
        ),
        CNavItem(
          title: const Text('Profile'),
          icon: SvgPicture.asset(
            "assets/icons/svg/outline/user.svg",
            color: Colors.amber,
          ),
          selectedIcon: SvgPicture.asset(
            "assets/icons/svg/solid/user.svg",
            color: Colors.pinkAccent,
          ),
        ),
      ],
      currentIndex: _currentIndex,
      onTap: (index) {
        setState(() {
          _currentIndex = index;
        });
      },
    );
```


## Example
Check `example app` for more details.

## Contribute
Issues and pull requests are welcomed!!

### Future Plans
- [x] Code format.
- [x] Make it more like native navigation bar in Flutter.
- [x] Better documentation.
- [ ] Optimizing Margin and Padding Implementation!
- [ ] More customizations!!
And more...

CHANGELOG.md

##[1.0.3]  - 2021-09-16
* `ReadMe.MD` updated.
* `Changelog.md` updated.
* Release Version updated to 1.0.3 from github by @princeteck in https://github.com/prodevplus/cnav/pull/5

##[1.0.2]  - 2021-09-16
* screenshot links updated with live likes from github by @princeteck in https://github.com/prodevplus/cnav/pull/4

##[1.0.1]  - 2021-09-16
* `Changelog.md` updated.

## [1.0.0] - 2021-09-16
* Task: Documentation updated.
* major improvements and new examples added for better understanding of implementation by @princeteck in https://github.com/prodevplus/cnav/pull/3

## [0.1.2] - 2021-03-14
* Task: Documentation updated.
* Task: Position between Normal and Selected Icon fixed.

## [0.1.1] - 2021-03-14
* Task: Documentation updated.

## [0.1.0] - 2021-03-13
* Task: Initial release.
* Task: License added.
* Task: Change log Updated.


lib/src/main.dart
// Copyright (c) 2021 Yuvraj Sonawane | PRODEVPLUS
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
import 'dart:ui';
import 'dart:math' as math;
import 'package:cnav/cnav.dart';
import 'package:cnav/src/utils/cnav_tile.dart';
import 'package:flutter/material.dart';
import 'package:cnav/src/utils/size_util.dart';
import 'package:cnav/src/animations/beacon_painter.dart';
import 'package:cnav/src/constants/ui/ui_constants.dart';

class CNav extends StatefulWidget {
  CNav({
    Key? key,
    required this.items,
    this.margin,
    this.padding,
    this.selectedColor,
    this.unSelectedColor,
    required this.onTap,
    this.currentIndex = 0,
    this.iconSize = 24.0,
    this.scaleFactor = 0.2,
    this.elevation = 8.0,
    this.borderRadius = Radius.zero,
    this.backgroundColor = Colors.white,
    this.strokeColor = Colors.blueAccent,
    this.bubbleCurve = Curves.linear,
    this.scaleCurve = Curves.linear,
    this.isFloating = false,
    this.blurEffect = false,
    this.opacity = 0.8,
    this.floatingPadding = 0.0,
  })  : assert(scaleFactor <= 0.5, 'Scale factor must smaller than 0.5'),
        assert(scaleFactor > 0, 'Scale factor must bigger than 0'),
        assert(0 <= currentIndex && currentIndex < items.length),
        super(key: key);

  ///
  /// scale factor for the icon scale animation effect.
  /// default is `0.2`.
  final double scaleFactor;

  ///
  /// boolean that control if navigation bar perform floating.
  /// default is false
  ///
  final bool isFloating;

  ///
  /// boolean that control if navigation bar perform floating.
  /// default is false
  ///
  final double floatingPadding;

  ///
  /// Border radius for navigation bar
  ///
  final Radius borderRadius;

  /// The z-coordinate of this [CustomNavigationBar].
  ///
  /// If null, defaults to `8.0`.
  final double elevation;

  ///
  /// Item data in [CustomNavigationBarItem]
  ///
  final List<CNavItem> items;

  ///
  /// [Color] when [CustomNavigationBarItem] is selected.
  ///
  /// default color is [blueAccent].
  final Color? selectedColor;

  ///
  /// [Color] when [CustomNavigationBarItem] is not selected.
  ///
  /// default color is [grey[600]].
  final Color? unSelectedColor;

  ///
  /// callback function when item tapped
  ///
  final Function(int) onTap;

  ///
  /// current index of navigation bar.
  ///
  final int currentIndex;

  ///
  /// size of icon.
  /// also represent the max radius of bubble effect animation.
  final double iconSize;

  ///
  /// Margin and Padding Properties.
  ///
  final EdgeInsetsGeometry? margin, padding;

  ///
  /// Background color of [CustomNavigationBar]
  ///
  final Color backgroundColor;

  ///
  /// stroke color.
  /// default is [Colors.blueAccent].
  final Color strokeColor;

  ///
  /// animation curve of bubble effect
  ///
  final Curve bubbleCurve;

  ///
  /// animation curve of scale effect
  ///
  final Curve scaleCurve;

  ///
  /// bool to control if navigation bar use blur effect
  /// default is [false]
  ///
  final bool blurEffect;

  ///
  /// When [blurEffect] is [true], control the opacity of navigation bar
  /// default is [0.8]
  ///
  final double opacity;

  @override
  _CNavState createState() => _CNavState();
}

class _CNavState extends State<CNav> with TickerProviderStateMixin {
  late List<double> _radiuses;
  late List<double> _sizes;
  late AnimationController _controller;
  late AnimationController _scaleController;

  double _bubbleRadius = 0.0;
  late double _maxRadius;

  double _itemPadding = 0.0;

  @override
  void initState() {
    super.initState();
    _bubbleRadius = 0.0;
    _radiuses = List<double>.generate(widget.items.length, (index) {
      return _bubbleRadius;
    });

    _sizes = List<double>.generate(widget.items.length, (index) {
      return 0.0;
    });
    _maxRadius = widget.iconSize;
    _controller = AnimationController(
      vsync: this,
      duration: Duration(milliseconds: 300),
    );
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  void didUpdateWidget(CNav oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (widget.currentIndex != oldWidget.currentIndex) {
      _startAnimation(widget.currentIndex);
      _startScale(widget.currentIndex);
    }
  }

  void _startAnimation(int index) {
    _controller = AnimationController(
      vsync: this,
      duration: Duration(milliseconds: 300),
    );
    CurvedAnimation _curvedAnimation = CurvedAnimation(
      parent: _controller,
      curve: widget.bubbleCurve,
    );
    Tween<double>(begin: 0.0, end: 1.0).animate(_curvedAnimation)
      ..addListener(() {
        setState(() {
          _radiuses[index] = _maxRadius * _curvedAnimation.value;
        });
      });
    _controller.forward();
  }

  void _startScale(int index) {
    _scaleController = AnimationController(
      vsync: this,
      duration: Duration(milliseconds: 400),
    )..addStatusListener((status) {
        if (status == AnimationStatus.completed) {
          _scaleController.reverse();
        }
      });
    CurvedAnimation _scaleAnimation = CurvedAnimation(
      parent: _scaleController,
      curve: widget.scaleCurve,
      reverseCurve: widget.scaleCurve.flipped,
    );

    Tween<double>(begin: 0.0, end: 1.0).animate(_scaleAnimation)
      ..addListener(() {
        setState(() {
          _sizes[index] = _scaleAnimation.value * widget.scaleFactor;
        });
      });
    _scaleController.forward();
  }

  Widget _buildLabel(int index) {
    // unselected
    if (index != widget.currentIndex) {
      if (widget.items[index].title == null && widget.isFloating) {
        return SizedBox();
      } else {
        return widget.items[index].title ?? SizedBox();
      }
    } else {
      //selected
      if (widget.isFloating && widget.items[index].title == null) {
        return SizedBox();
      } else {
        return widget.items[index].selectedTitle ?? SizedBox();
      }
    }
  }

  Widget _buildIcon(int index) {
    return FittedBox(
      child: Container(
        margin: widget.margin ?? EdgeInsets.all(10),
        height: ((widget.items[index].icon is Icon)
                ? (widget.items[index].icon as Icon).size
                : null) ??
            widget.iconSize,
        width: ((widget.items[index].icon is Icon)
                ? (widget.items[index].icon as Icon).size
                : null) ??
            widget.iconSize,
        child: CustomPaint(
          painter: BeaconPainter(
            color: widget.strokeColor,
            beaconRadius: ((widget.items[index].icon is Icon)
                        ? (widget.items[index].icon as Icon).size
                        : null) !=
                    null
                ? _radiuses[index] * 1.5
                : _radiuses[index],
            maxRadius: ((widget.items[index].icon is Icon)
                        ? (widget.items[index].icon as Icon).size
                        : null) !=
                    null
                ? _maxRadius * 1.5
                : _maxRadius,
            offset: Offset(
              (((widget.items[index].icon is Icon)
                          ? (widget.items[index].icon as Icon).size
                          : null) ??
                      widget.iconSize) /
                  2,
              (((widget.items[index].icon is Icon)
                          ? (widget.items[index].icon as Icon).size
                          : null) ??
                      widget.iconSize) /
                  2,
            ),
          ),
          child: FittedBox(
            child: Padding(
              padding: widget.padding ?? EdgeInsets.zero,
              child: CNavTile(
                iconSize: (((widget.items[index].icon is Icon)
                        ? (widget.items[index].icon as Icon).size
                        : null) ??
                    widget.iconSize),
                scale: _sizes[index],
                selected: index == widget.currentIndex,
                item: widget.items[index],
                selectedColor: widget.selectedColor ?? KUi.colors.defaultColor,
                unSelectedColor:
                    widget.unSelectedColor ?? KUi.colors.defaultUnSelectedColor,
                iconPadding: _itemPadding,
              ),
            ),
          ),
        ),
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    SizeUtil().init(context);
    final double additionalBottomPadding =
        math.max(MediaQuery.of(context).padding.bottom, 0.0);

    final height = KUi.measurements.defaultHeight +
        (widget.isFloating ? 0.0 : additionalBottomPadding);

    _itemPadding = (MediaQuery.of(context).size.width -
            widget.items.length * widget.iconSize) /
        (widget.items.length * 2);

    if (widget.isFloating) {
      _itemPadding = (MediaQuery.of(context).size.width -
              widget.items.length * widget.iconSize -
              32) /
          (widget.items.length * 2);
    }

    Widget bar = Container(
      margin: widget.margin ?? EdgeInsets.zero,
      height: height,
      width: SizeUtil.screenWidth,
      child: Material(
        color: widget.backgroundColor,
        elevation: widget.elevation,
        borderRadius: BorderRadius.all(widget.borderRadius),
        child: Row(
          mainAxisAlignment: MainAxisAlignment.spaceAround,
          crossAxisAlignment: CrossAxisAlignment.center,
          children: List<Widget>.generate(
            widget.items.length,
            (i) => GestureDetector(
              behavior: HitTestBehavior.opaque,
              onTap: () {
                widget.onTap(i);
              },
              child: Center(
                child: FittedBox(
                  child: Column(
                    mainAxisAlignment: MainAxisAlignment.center,
                    crossAxisAlignment: CrossAxisAlignment.center,
                    children: [
                      (widget.items[i].title != null)
                          ? Transform.scale(scale: 1.5, child: _buildIcon(i))
                          : _buildIcon(i),
                      _buildLabel(i),
                      (widget.items[i].title != null)
                          ? SizedBox(height: 10)
                          : SizedBox(),
                    ],
                  ),
                ),
              ),
            ),
          ).toList(),
        ),
      ),
    );

    if (widget.blurEffect) {
      return Padding(
        padding: widget.isFloating
            ? EdgeInsets.only(
                left: getProportionateScreenWidth(widget.floatingPadding),
                right: getProportionateScreenWidth(widget.floatingPadding),
                top: 0,
                bottom: additionalBottomPadding,
              )
            : EdgeInsets.zero,
        child: ClipRRect(
          borderRadius: BorderRadius.all(widget.borderRadius),
          child: BackdropFilter(
            filter: ImageFilter.blur(
              sigmaX: 5.0,
              sigmaY: 10.0,
            ),
            child: Opacity(
              opacity: 0.6,
              child: bar,
            ),
          ),
        ),
      );
    } else {
      return Padding(
        padding: widget.isFloating
            ? EdgeInsets.only(
                left: getProportionateScreenWidth(widget.floatingPadding),
                right: getProportionateScreenWidth(widget.floatingPadding),
                top: 0,
                bottom: additionalBottomPadding,
              )
            : EdgeInsets.zero,
        child: bar,
      );
    }
  }
}
